Mathlib/AlgebraicTopology/SimplicialSet/Path.lean

```diff
@@ -1,145 +1,294 @@
   arrow_src i := by
-    dsimp [SimplicialObject.δ]
-    simp only [← FunctorToTypes.map_comp_apply, ← op_comp]
-    rw [SimplexCategory.δ_one_mkOfSucc]
-    simp only [len_mk, Fin.coe_castSucc, Fin.coe_eq_castSucc]
+    dsimp only [tr, trunc, SimplicialObject.Truncated.trunc, incl,
+      whiskeringLeft_obj_obj, id_eq, Functor.comp_map, Functor.op_map,
+      Quiver.Hom.unop_op]
+    rw [← FunctorToTypes.map_comp_apply, ← op_comp, ← tr_comp, δ_one_mkOfSucc,
+      Fin.coe_castSucc, Fin.coe_eq_castSucc]
   arrow_tgt i := by
-    dsimp [SimplicialObject.δ]
-    simp only [← FunctorToTypes.map_comp_apply, ← op_comp]
-    rw [SimplexCategory.δ_zero_mkOfSucc]
-
-lemma spine_map_vertex {n : ℕ} (x : X _[n]) {m : ℕ} (φ : ([m] : SimplexCategory) ⟶ [n])
-    (i : Fin (m + 1)) :
-    (spine X m (X.map φ.op x)).vertex i = (spine X n x).vertex (φ.toOrderHom i) := by
-  dsimp [spine]
-  rw [← FunctorToTypes.map_comp_apply]
+    dsimp only [tr, trunc, SimplicialObject.Truncated.trunc, incl,
+      whiskeringLeft_obj_obj, id_eq, Functor.comp_map, Functor.op_map,
+      Quiver.Hom.unop_op]
+    rw [← FunctorToTypes.map_comp_apply, ← op_comp, ← tr_comp, δ_zero_mkOfSucc]
+
+/-- Further truncating `X` above `m` does not change the `m`-spine. -/
+lemma trunc_spine (k m : ℕ) (h : m ≤ k + 1) (hₙ : k ≤ n) :
+    ((trunc (n + 1) (k + 1)).obj X).spine m = X.spine m :=
   rfl
 
-lemma spine_map_subinterval {n : ℕ} (j l : ℕ) (hjl : j + l ≤ n) (Δ : X _[n]) :
-    X.spine l (X.map (subinterval j l (by omega)).op Δ) =
-      (X.spine n Δ).interval j l (by omega) := by
+variable (m : ℕ) (hₘ : m ≤ n + 1)
+
+lemma spine_map_vertex (Δ : X.obj (op ⟨.mk m, hₘ⟩)) (a : ℕ) (hₐ : a ≤ n + 1)
+    (φ : (⟨.mk a, hₐ⟩ : SimplexCategory.Truncated (n + 1)) ⟶ ⟨.mk m, hₘ⟩)
+    (i : Fin (a + 1)) :
+    (X.spine a hₐ (X.map φ.op Δ)).vertex i =
+      (X.spine m hₘ Δ).vertex (φ.toOrderHom i) := by
+  dsimp only [spine_vertex]
+  rw [← FunctorToTypes.map_comp_apply, ← op_comp, ← tr_comp,
+    SimplexCategory.const_comp]
+
+lemma spine_map_subinterval (j l : ℕ) (h : j + l ≤ m)
+    (Δ : X.obj (op ⟨.mk m, hₘ⟩)) :
+    X.spine l (by omega) (X.map (tr (subinterval j l h)).op Δ) =
+      (X.spine m hₘ Δ).interval j l h := by
   ext i
-  · simp only [spine_vertex, Path.interval, ← FunctorToTypes.map_comp_apply, ← op_comp,
+  · dsimp only [spine_vertex, Path.interval]
+    rw [← FunctorToTypes.map_comp_apply, ← op_comp, ← tr_comp,
       const_subinterval_eq]
-  · simp only [spine_arrow, Path.interval, ← FunctorToTypes.map_comp_apply, ← op_comp,
+  · dsimp only [spine_arrow, Path.interval]
+    rw [← FunctorToTypes.map_comp_apply, ← op_comp, ← tr_comp,
       mkOfSucc_subinterval_eq]
 
-/-- Two paths of the same nonzero length are equal if all of their arrows are equal. -/
+end Truncated
+
+/-- A path of length `n` in a simplicial set `X` is defined as a 1-truncated
+path on the 1-truncation of `X`. -/
+def Path (X : SSet.{u}) (n : ℕ) := truncation 1 |>.obj X |>.Path n
```

Comment:
Because `Path` is defined here as `truncation 1 |>.obj X |>.Path n` rather than using `Path₁` directly, it unfolds to `((Truncated.trunc 1 1).obj ((truncation 1).obj X)).Path₁ n`. 

While this is defeq to `((truncation 1).obj X).Path₁ n`, it is at the very least confusing to see the extra layer of truncation in the infoview. 

Unfortunately, the only obvious solution I see is to define `SSet.Path` directly in terms of `Path₁`, which I think means duplicating all of the `Truncated.Path` constructions on `Path₁`, as was done before @joelriou suggested cutting down on this duplication.

Is there a better solution (and is it worth implementing)?

Reply:
We should not have to see any `Path₁` in the infoview. Once a reasonable API is developed for `Truncated.Path` (I think you should add abbreviations `Truncated.Path.vertex/arrow` similarly as the definitions you have done for `SSet.Path`), you may define `SSet.Path` as an abbrev (for `Truncated.Path` of the `1`-truncation). Then, we can add abbreviations `SSet.Path.vertex/arrow`, so that `vertex` takes its values in `X _[0]`. Some API definitions and lemmas will to be duplicated, but the point is that there is only one underlying structure, and all the definitions/proofs for `SSet` should be one-liner applications of the corresponding API for `Truncated.SSet.Path`.

Reply:
Thanks. I believe I have implemented this in [149cb27](https://github.com/leanprover-community/mathlib4/pull/20668/commits/149cb276b2d1a575ed0b698e7b3776ab6beff0b0), and everything seemed to work out well.

Reply:
Well, everything except the simpNF linter.

Reply:
Maybe the `spine` should be a def rather than abbrev?

Reply:
I have changed `spine` and `map` to be defs at the `SSet` level, which allowed tagging them `@[simp]` again. I also changed `interval` to be a def, because it seems to fit in the same category as `spine` and `map`. At the `SSet` level, we now have the following:
- `abbrev Path`
- `abbrev vertex`
- `abbrev arrow`
- `def interval`
- `def map`
- `def spine`

My thought is that this should be the right mix of reducibility to keep `dsimp` from dropping down to the `SSet.Truncated` level when you are trying to work at the `SSet` level. But, it may still need some tweaking.

Reply:
You can see in [`bebab82`](https://github.com/leanprover-community/mathlib4/pull/20668/commits/bebab8276d34aee450b924fcd0cdcf9ea8f08dfd) that the change mentioned above allowed removing some extra `dsimp` truncation lemmas from one spot where I don't think we should need to deal with truncation. I do wish I had some more examples of this to experiment with, though.

Reply:
Another thought -- Maybe `Path₁` should be private? We would need to add constructors `SSet.Path.mk` and `SSet.Truncated.Path.mk`, but it might make it easier to identify the places where the API falls short.